fvm/evm/stdlib/contract.go

```diff
@@ -38,6 +40,272 @@ var evmAddressBytesType = sema.NewConstantSizedType(nil, sema.UInt8Type, types.A
+			TypeAnnotation: sema.NewTypeAnnotation(sema.ByteArrayType),
+		},
+	},
+	ReturnTypeAnnotation: sema.NewTypeAnnotation(
+		sema.NewVariableSizedType(nil, sema.AnyStructType),
+	),
+}
+
+func newInternalEVMTypeDecodeABIFunction(
+	gauge common.MemoryGauge,
+) *interpreter.HostFunctionValue {
+	return interpreter.NewHostFunctionValue(
+		gauge,
+		internalEVMTypeDecodeABIFunctionType,
+		func(invocation interpreter.Invocation) interpreter.Value {
+			inter := invocation.Interpreter
+			locationRange := invocation.LocationRange
+
+			// Get `types` argument
+
+			typesArray, ok := invocation.Arguments[0].(*interpreter.ArrayValue)
+			if !ok {
+				panic(errors.NewUnreachableError())
+			}
+
+			// Get `data` argument
+
+			dataValue, ok := invocation.Arguments[1].(*interpreter.ArrayValue)
+			if !ok {
+				panic(errors.NewUnreachableError())
+			}
+
+			data, err := interpreter.ByteArrayValueToByteSlice(inter, dataValue, locationRange)
+			if err != nil {
+				panic(err)
+			}
+
+			var arguments gethABI.Arguments
+			for i := 0; i < typesArray.Count(); i++ {
+				arg := typesArray.Get(inter, locationRange, i)
+				switch arg.String() {
+				case "Type<String>()":
+					typ, err := gethABI.NewType("string", "", nil)
+					if err != nil {
+						panic(err)
+					}
+					arguments = append(arguments, gethABI.Argument{Type: typ})
+				case "Type<UInt64>()":
+					typ, err := gethABI.NewType("uint64", "", nil)
+					if err != nil {
+						panic(err)
+					}
+					arguments = append(arguments, gethABI.Argument{Type: typ})
+				case "Type<Bool>()":
+					typ, err := gethABI.NewType("bool", "", nil)
+					if err != nil {
+						panic(err)
+					}
+					arguments = append(arguments, gethABI.Argument{Type: typ})
+				case "Type<A.0000000000000001.EVM.EVMAddress>()":
+					typ, err := gethABI.NewType("address", "", nil)
+					if err != nil {
+						panic(err)
+					}
+					arguments = append(arguments, gethABI.Argument{Type: typ})
+				case "Type<[UInt64]>()":
+					typ, err := gethABI.NewType("uint64[]", "", nil)
+					if err != nil {
+						panic(err)
+					}
+					arguments = append(arguments, gethABI.Argument{Type: typ})
+				}
+			}
+
+			decoded, err := arguments.Unpack(data)
+			if err != nil {
+				panic(err)
+			}
+
+			values := make([]interpreter.Value, 0)
+			for _, arg := range decoded {
+				switch value := arg.(type) {
+				case string:
+					values = append(values, interpreter.NewStringValue(
+						inter,
+						common.NewStringMemoryUsage(len(value)),
+						func() string {
+							return value
+						},
+					))
+				case uint64:
+					values = append(values, interpreter.NewUInt64Value(
+						inter,
+						func() uint64 {
+							return value
+						},
+					))
+				case bool:
+					values = append(values, interpreter.BoolValue(value))
+				case gethCommon.Address:
+					var address types.Address
+					copy(address[:], value.Bytes())
+					values = append(values, EVMAddressToAddressBytesArrayValue(inter, address))
```

Comment:
Here I should actually return the whole `EVM.EVMAddress` struct, and not just its `bytes`. I didn't find a way how to create it though.